src/inmanta/plugins.py

```diff
@@ -341,6 +353,63 @@ def to_dsl_type(python_type: type[object]) -> inmanta_type.Type:
     return inmanta_type.Type()
 
 
+def validate_and_convert_to_python_domain(expected_type: inmanta_type.Type, value: object) -> object:
+    """
+    Given a model domain value and an inmanta type, produce the corresponding python object
+
+    Unknowns are not handled by this method!
+    """
+    expected_type.validate(value)
+
+    if isinstance(value, NoneValue):
+        # if the type is not nullable, it will fail when validating
+        # if the value is None, it becomes None
+        return None
+
+    if expected_type.has_custom_to_python():
+        return expected_type.to_python(value)
+
+    return DynamicProxy.return_value(value)
+
+
+def primitive_python_type_to_model_domain(intype: type) -> inmanta_type.Type:
```

Comment:
Any reason for this custom implementation rather than `to_dsl_type(intype).type_string()`, raising an error if `type_string()` returns None?

Reply:
that is indeed much better.